src/borg/item.pyx

```diff
@@ -359,60 +377,127 @@ class ManifestItem(PropDict):
     config = PropDict._make_property('config', dict)
     item_keys = PropDict._make_property('item_keys', tuple)
 
-    def compare_link(item1, item2):
-        # These are the simple link cases. For special cases, e.g. if a
-        # regular file is replaced with a link or vice versa, it is
-        # indicated in compare_mode instead.
-        if item1.get('deleted'):
+class ItemDiff:
+    """
+    Comparison of two items from different archives.
+
+    The items may have different paths and still be considered equal (e.g. for renames).
+    It does not include extended or time attributes in the comparison.
+    """
+
+    def __init__(self, item1, item2, chunk_iterator1, chunk_iterator2, numeric_owner=False, can_compare_chunk_ids=False):
+        self.item1 = item1
+        self.item2 = item2
+        self._numeric_owner = numeric_owner
+        self.equal = self._equal(chunk_iterator1, chunk_iterator2, can_compare_chunk_ids)
+
+    def __repr__(self):
+        if self.equal:
+            return 'equal'
+
+        changes = []
+
+        if self.item1.is_link() or self.item2.is_link():
+            changes.append(self._link_string())
+
+        if 'chunks' in self.item1 and 'chunks' in self.item2:
+            changes.append(self._content_string())
+
+        if self.item1.is_dir() or self.item2.is_dir():
+            changes.append(self._dir_string())
+
+        if not (self.item1.get('deleted') or self.item2.get('deleted')):
+            changes.append(self._owner_string())
+            changes.append(self._mode_string())
+
+        return ' '.join((x for x in changes if x))
+
+    def _equal(self, chunk_iterator1, chunk_iterator2, can_compare_chunk_ids):
+        # if both are deleted, there is nothing at path regardless of what was deleted
+        if self.item1.get('deleted') and self.item2.get('deleted'):
```

Comment:
it can be solved in other ways, but i guess the point was to have an easy comparison without introducing a lot of special cases in the comparison function.

e.g. the deleted_item has chunks = [], so if it is compared to a normal item, all chunks show up as added/removed naturally. Also, size increase/decrease will also just work. Same thing can be done for all sorts of metadata, just put an empty/zero value in deleted_item.